synapse/lib/stormlib/aha.py

```diff
@@ -531,5 +620,223 @@ async def _methPoolSvcDel(self, svcname):
             }
         }
         '''
-    }
+    },
+    {
+        'name': 'aha.svc.mirror',
+        'descr': '''Query the AHA services and their mirror relationships.
+
+Note: non-mirror services are not displayed.
+        ''',
+        'cmdargs': (
+            ('--timeout', {'help': 'The number of seconds to wait for the mirrors to sync.',
+                           'default': 60, 'action': 'store'}),
+        ),
+        'storm': '''
+        init {
+            $conf = ({
+                "columns": [
+                    {"name": "name", "width": 40},
+                    {"name": "role", "width": 9},
+                    {"name": "online", "width": 7},
+                    {"name": "ready", "width": 6},
+                    {"name": "host", "width": 16},
+                    {"name": "port", "width": 8},
+                    {"name": "version", "width": 12},
+                    {"name": "nexus idx", "width": 10},
+                ],
+                "separators": {
+                    "row:outline": false,
+                    "column:outline": false,
+                    "header:row": "#",
+                    "data:row": "",
+                    "column": "",
+                },
+            })
+            $printer = $lib.tabular.printer($conf)
+            $timeout = $cmdopts.timeout
+        }
+
+        function get_cell_infos(vname) {
+            $cell_infos = ({})
+            for $info in $lib.aha.callPeerApi($vname, getCellInfo) {
+                $svcname = $info.0
+                ($ok, $info) = $info.1
+                if $ok {
+                    $cell_infos.$svcname = $info
+                }
+            }
+            return($cell_infos)
+        }
+
+        function build_status_list(members, cell_infos) {
+            $group_status = $lib.list()
+            for $svc in $members {
+                $svcinfo = $svc.svcinfo
+                $svcname = $svc.name
+                $status = ({
+                    'name': $svcname,
+                    'role': '<unknown>',
+                    'online': $lib.dict.has($svcinfo, 'online'),
```

Comment:
Is this correct? If the `online` key is in the dictionary then it's online? What's the value of this key? Could there be a `$svcinfo.online = (false)`?

Reply:
I'll double check but from what I remember if the service is offline the key is gone.

Reply:
Yep, it's correct. The `online` key is a session ID of the socket associated with the connection that the service has made to AHA so that they can be safely check+pop'd without a race condition for the next session being established before we realize the socket has gone down. It will either be a `guid` or not populated. Mapping it's presence to a bool is correct here :+1: